Connect to Postgres (#281)
xataio/mdx-docs: 020-Connect/010-Postgres/001-connect.mdx

---
title: Connect to Postgres
navTitle: Connect
keywords: ['postgres', 'connect', 'dsn']
description: Learn how to connect to Postgres
slug: postgres/connect
published: true
---

> <Alert status="info">The Postgres wire protocol is still considered BaseAudioContext.</Alert>

Reviewer: I guess this was intended to be Beta
```suggestion
<Alert status="info">The Postgres wire protocol is still considered Beta.</Alert>
```

This page shows you how to connect to Postgres, by explaining how a connection string is built, the different parameters and examples to connect via [psql](https://www.postgresql.org/docs/current/app-psql.html).

> 

Reviewer: probably worth adding some more intro text about Xata allowing direct access to underlying postgres, but this is a good start

Author: 100%, I want to get the page outline first and add more meat to the bone.

Author: Intro text in place

 PostgreSQL 15.4 on aarch64-unknown-linux-gnu, compiled by aarch64-unknown-linux-gnu-gcc (GCC) 9.5.0, 64-bit
(1 row)
```

### Shared cluster

Version: `15.4`

In a shared cluster, you operate in a multi tenant environment, sharing the same Postgres version. Minor version upgrades are continuously updated.
If you want control of the version, you should consider a dedicated cluster.

### Dedicated cluster

You can select which version you want, currently supported versions in all [available regions](/docs/getting-started/available-regions):

- 15.4
- 15.5
- 14.10
- 14.9

You can control the maintenance windows and the version upgrades yourself.

## Extensions

The available, out-of-the-box extensions are:

- list

If you are on a dedicated cluster you can install extensions yourself, please consult the [Postgres documentation](https://www.postgresql.org/docs/current/sql-createextension.html) for a how to.

## Supported Statements

Xata offers different approaches to Postgres, from SQL over HTTP to pure Postgres over the wire protocol to a shared or dedicated cluster.
Each come with a security scope, to provide isolation in a multi tenant environment.

> |                   | FUNCTIONs | DQL statements | (some) DML statements | DDL statements |

Reviewer 2: ```suggestion
|                   | FUNCTIONS| DQL statements | (some) DML statements | DDL statements |
```

| ----------------- | --------- | -------------- | --------------------- | -------------- |
| SQL over HTTP     | N         | Y              | Y                     | Y              |
| Shared cluster    | N         | Y              | Y                     | Y              |
| Dedicated cluster | Y         | Y              | Y                     | Y              |

> The table below, is providing a more granularity picture of the supported statements.

Reviewer 2: ```suggestion
The table below, is providing a more granular picture of the supported statements.
```

> |                         | SQL over HTTP | Shared Cluster | Dedicated Cluster |

Reviewer 2: The distinction of SQL over HTTP and Shared/Dedicated cluster isn't very clear, I suppose this means wire protocol? Maybe rename to `Wire on Shared Cluster` and `Wire on Dedicated Cluster`

Author: Funnily enough, I was thinking exactly the same as I was curating the table. I tried to address it in

> Xata offers different approaches to Postgres, from SQL over HTTP to pure Postgres over the wire protocol to a shared or dedicated cluster.

But, likely needs more of an explicit call-out. Thanks Kostas

Author: addressed

| ----------------------- | ------------- | -------------- | ----------------- |
| `SELECT`                | Y             | Y              | Y                 |
| `INSERT`                | Y             | Y              | Y                 |
| `DELETE`                | Y             | Y              | Y                 |
| `UPDATE`                | Y             | Y              | Y                 |
| `SHOW`                  | Y             | Y              | Y                 |
| `ALTER TABLE`           | Y             | Y              | Y                 |
| `CREATE TABLE`          | Y             | Y              | Y                 |
| `DROP TABLE`            | Y             | Y              | Y                 |
| `BEGIN`                 | N             | Y              | Y                 |
| `COMMIT`                | N             | Y              | Y                 |

> | `ROLLBACK`              | Y             | Y              | Y                 |

Reviewer 3: ```suggestion
| `ROLLBACK`              | N             | Y              | Y                 |
```

> | `SET`                   | Y             | Y              | Y                 |

Reviewer 3: ```suggestion
| `SET`                   | N             | Y              | Y                 |
```

> | `COPY`                  | Y             | Y              | Y                 |

Reviewer 3: ```suggestion
| `COPY`                  | N             | Y              | Y                 |
```

| `LOCK`                  | Y             | Y              | Y                 |
| `PREPARE`               | Y             | Y              | Y                 |

> | `EXECUTE`               | Y             | Y              | Y                 |

Reviewer 3: ```suggestion
| `EXECUTE`               | N             | Y              | Y                 |
```

| `TRUNCATE`              | Y             | Y              | Y                 |

> | `ALTER TABLESPACE`      | N             | N              | Y                 |

Reviewer 3: Also, I decided to remove support for tablespaces for now. It would allow users to change the path where the tables are stored. As they do not have access to the machine Postgresql is running on, it does not make sense for them to configure it.

Author: addressed via https://github.com/xataio/mdx-docs/pull/281/commits/b89bc30ca0617bd5b704e334161a4cf411ed9df7 - @kvch could you confirm this is correct and resolve the conversation? thanks

| `ALTER DOMAIN`          | N             | N              | Y                 |
| `CREATE DOMAIN`         | N             | N              | Y                 |
| `DROP DOMAIN`           | N             | N              | Y                 |
| `ALTER FUNCTION`        | N             | N              | Y                 |
| `CREATE FUNCTION`       | N             | N              | Y                 |
| `DROP FUNCTION`         | N             | N              | Y                 |
| `EXPLAIN`               | N             | N              | Y                 |
| `REINDEX`               | N             | N              | Y                 |
| `ALTER OPERATOR CLASS`  | N             | N              | Y                 |
| `CREATE OPERATOR CLASS` | N             | N              | Y                 |
| `DROP OPERATOR CLASS`   | N             | N              | Y                 |
| `CREATE EXTENSION`      | N             | N              | Y                 |
| `DROP EXTENSION`        | N             | N              | Y                 |
| `ALTER EXTENSION`       | N             | N              | Y                 |

> | Temporary tables        | N             | N              | Y                 |

Reviewer 3: Maybe `CREATE TEMPORARY TABLE` is more consistent than "Temporary tables". WDYT?

Author: resolved via https://github.com/xataio/mdx-docs/pull/281/commits/fc845dc8671ee7763ea431d68c2ccc45d690f4e0

> | Temporary tables        | N             | N              | Y                 |

Author: https://github.com/xataio/mdx-docs/pull/281/files#r1500637177 --> I think yes :)

## Limitations

### Unsupported statements and functions

The following statements and/or functions are not available:

- `SELECT user;`

> - `SHOW password_encryption;`

Reviewer 3: We should also include that we do not support changing the roles. Example commands `SET ROLE`, `RESET ROLE` and `SET SESSION AUTHORIZATION`.

Author: resolved via https://github.com/xataio/mdx-docs/pull/281/commits/fc845dc8671ee7763ea431d68c2ccc45d690f4e0

### Connection limits

- SQL over HTTP to a shared cluster has the rate limits in place, for a dedicated cluster there are none.
- Shared cluster - TODO, amount of available connections
- Dedicated cluster - no limit to connections

### `xata_` Column prefix

Xata reserves the `xata_` column prefix for [special columns](docs/concepts/data-model#special-columns) and rejects any custom created columns with the prefix.

### sslmode

Xata does not allow plain text connections, this may have an effect on the client you are using and the default that is inferred by such.

> If you face an issue, be append the SSL mode `require` option to the connection string or pass in a parameterized way depending on the tool you are using.

Reviewer 4: I would skip the appending `require` suggestion since in `psql` the default behaviour is attempting SSL, so this wouldn't make any difference if the user is facing a connectivity issue.

However, I would provide some guidelines about which mode to use in `psql`, something like:

Xata recommends that the client verifies the server certificate, in `psql` this would be achieved with the `verify-ca` and `verify-full` modes. We will need to provide a certificate store to do this, it is usually present in most operative systems, for instance:
- MacOS: /etc/ssl/cert.pem
- GNU/Linux: /etc/ssl/certs/ca-certificates.crt

Using `psql` on Mac to establish a fully secure connection to Xata is as easy as:

```
psql "sslmode=verify-full sslrootcert=/etc/ssl/cert.pem user= ws1234 password=xau_apikey123456 host=us-east-1.sql.xata.sh dbname=Games:main"
```

Keep in mind that different clients will provide different configuration options.

Author: Thank you for the extended feedback Alejandro, I'll address and ping you.

Author: first draft via https://github.com/xataio/mdx-docs/pull/281/commits/5e0b38b19a8f156e0768e82e6e19308a7393523a

---
title: Connect to Postgres
navTitle: Connect
keywords: ['postgres', 'connect', 'dsn']
description: Learn how to connect to Postgres
slug: postgres/connect
published: true
status: beta
---

This page shows you how to connect to Postgres, by explaining how a connection string is built, the different parameters and examples to connect via [psql](https://www.postgresql.org/docs/current/app-psql.html).

## Connection string

> The connection string to connect to Postgres is constructed the following:

Reviewer 5: ```suggestion
The connection string to connect to Postgres needs to be in the following format:
```

```
postgresql://[User]:[Password]@[Host]:[Port]/[Database]
```

The following conversion table illustrates how the connection string parameters are mapped to Xata.

| Postgres | Xata                       | Example                 |
| -------- | -------------------------- | ----------------------- |
| User     | Workspace Id               | `ws1234`                |
| Password | API Key                    | `xau_apikey123456`      |
| Host     | Region plus `.sql.xata.sh` | `us-east-1.sql.xata.sh` |
| Port     | `5432` or empty            | `5432`                  |
| Database | Database and branch name   | `Games:main`            |

With the example values from above, the connection string to connect to our database `Games` on branch `main` is constructed as:

```
postgresql://ws1234:xau_apikey123456@us-east-1.sql.xata.sh:5432/Games:main
```

> The connection strings for the HTTP access and for the Postgres wire protocol are shown to you when you create a new database, afterwards you can always navigate to the database settings to see the URL construction.

Reviewer 5: ```suggestion
The connection strings for the HTTP access and for the Postgres wire protocol are shown to you when you create a new database. You can always see how the connection string is formed later on by navigating to the database settings page.
```

![Database settings showing connection strings](images/connection-string.png)

To create a new API key, or manage your existing ones, you need to open your [account settings](https://app.xata.io/settings).

Using psql to connect to your database, based off the previously constructed connection string with the example values:

```bash
psql "postgresql://ws1234:xau_apikey123456@us-east-1.sql.xata.sh:5432/Games:main"

Games:main=>
```

> ## Supported Versions

Reviewer 5: ```suggestion
## Supported versions
```
Lowercase to be consistent with the other titles

You can check the Postgres version with the following command:

```bash
psql "postgresql://[Workspace Id]:[API Key]@[Region].sql.xata.sh:[Port]/[Database]:[Branch]" -c "select version()"
```

```sql
                                                   version
-------------------------------------------------------------------------------------------------------------
 PostgreSQL 15.4 on aarch64-unknown-linux-gnu, compiled by aarch64-unknown-linux-gnu-gcc (GCC) 9.5.0, 64-bit
(1 row)
```

### Shared cluster

Version: `15.4`

In a shared cluster, you operate in a multi tenant environment, sharing the same Postgres version. Minor version upgrades are continuously updated.
If you want control of the version, you should consider a dedicated cluster.

### Dedicated cluster

> You can select which version you want, currently supported versions in all [available regions](/docs/getting-started/available-regions):

Reviewer 5: ```suggestion
In a dedicated cluster, you can select which version you want. The currently supported versions in all [available regions](/docs/getting-started/available-regions) are:
```

- 15.4
- 15.5
- 14.10
- 14.9

You can control the maintenance windows and the version upgrades yourself.

## Extensions

The available, out-of-the-box extensions are:

- list

If you are on a dedicated cluster you can install extensions yourself, please consult the [Postgres documentation](https://www.postgresql.org/docs/current/sql-createextension.html) for a how to.

## pg_dump and pg_restore

TODO

> ## Supported Statements

Reviewer 5: ```suggestion
## Supported statements
```
Lowercase to be consistent with the other titles

Xata offers different approaches to Postgres, from SQL over HTTP to pure Postgres over the wire protocol to a shared or dedicated cluster.

> Each come with a security scope, to provide isolation in a multi tenant environment.

Reviewer 5: ```suggestion
Each come with a security scope to provide isolation in a multi tenant environment.
```

|                   | FUNCTIONS | DQL statements | (some) DML statements | DDL statements |
| ----------------- | --------- | -------------- | --------------------- | -------------- |
| SQL over HTTP     | N         | Y              | Y                     | Y              |
| Shared cluster    | N         | Y              | Y                     | Y              |
| Dedicated cluster | Y         | Y              | Y                     | Y              |

> The table below, is providing a more granular picture of the supported statements.

Reviewer 5: ```suggestion
The table below provides a more granular picture of the supported statements.
```

| `DROP FUNCTION`          | N             | N              | Y                 |
| `EXPLAIN`                | N             | N              | Y                 |
| `REINDEX`                | N             | N              | Y                 |
| `ALTER OPERATOR CLASS`   | N             | N              | Y                 |
| `CREATE OPERATOR CLASS`  | N             | N              | Y                 |
| `DROP OPERATOR CLASS`    | N             | N              | Y                 |
| `CREATE EXTENSION`       | N             | N              | Y                 |
| `DROP EXTENSION`         | N             | N              | Y                 |
| `ALTER EXTENSION`        | N             | N              | Y                 |
| `CREATE TEMPORARY TABLE` | N             | N              | Y                 |

## Limitations

### Unsupported statements and functions

The following statements and/or functions are not available:

- `SELECT user`
- `SHOW password_encryption`
- `SET ROLE`
- `RESET ROLE`
- `SET SESSION AUTHORIZATION`
- `ALTER TABLESPACE`
- `CREATE TABLESPACE`
- `DROP TABLESPACE`

### Connection limits

> - SQL over HTTP to a shared cluster has the rate limits in place, for a dedicated cluster there are none.

Reviewer 5: ```suggestion
- SQL over HTTP to a shared cluster has rate limits in place. SQL over HTTP to a dedicated cluster does not have rate limits in place.
```

- Shared cluster - TODO, amount of available connections
- Dedicated cluster - no limit to connections

### `xata_` Column prefix

Xata reserves the `xata_` column prefix for [special columns](docs/concepts/data-model#special-columns) and rejects any custom created columns with the prefix.

### sslmode

Xata does not allow plain text connections, this may have an effect on the client you are using and the default that is inferred by such.

> If you face an issue with an error message as shown below, you will need to explicitly force SSL, by adding the sslmode to `require` in the connection string or pass in a parameterized way depending on the tool you are using.

Reviewer 5: ```suggestion
If you face an issue with an error message like the one shown below, you will need to explicitly force SSL, by setting the sslmode to `require` in the connection string or passing this setting in via a parameter depending on the tool you are using.
```

Reviewer 5: Might benefit from showing an example of a connection string with the param: `?sslmode=require` or however it is constructed.

Author: Added example via https://github.com/xataio/mdx-docs/pull/281/commits/cafbafcd3c7e2e0a8f8cb2c3f22ea8076806dba0

and rewording via https://github.com/xataio/mdx-docs/pull/281/commits/3a77ac892ae0941bbcb181ccb72e070a198dd024 (GH UI refused to let me commit the suggestion 🤷 )
